src/index.ts

```diff
@@ -1,31 +1,82 @@
-// Very thin implementation. Streamline.js does all the work.
-import { _ } from 'streamline-runtime';
 const fibers = require('fibers');
 
 export type Callback<T> = (err: any, result?: T) => void;
 export type Thunk<T> = (cb: Callback<T>) => void;
 
-export let wait = <T>(arg: Promise<T> | Thunk<T>): T => {
-	if (typeof arg === 'function') {
-		// fiberized test optimizes calls to streamlined thunks: (_: _) => T 
-		const fiberized = (arg as any)['fiberized-0'];
-		if (fiberized) return fiberized(true);
-		else return wait(_.promise(_ => _.cast<T>(arg as any)(_)));
+///
+/// ## run/wait
+/// * `promise = run(() => { wait(promise/callback); ... })`
+///    create a coroutine to write asynchronous code in a synchronous way.
+///
+/// All the job is done by fibers library.
+/// Those two functions are the core, others are goodies:
+///   * `promise = run(fn)` create a coroutine.
+///     This start a fiber, which is stopped when `fn` returns.
+///   * `result = wait(promise/callback)` encapsulate promise or callback.
+///     Concretely, the fiber is suspended while the asynchronous task is not finished, then it resumes.
+///     As many `wait()` as needed may be used in a run.
+export let wait = <T = any>(promiseOrCallback: Promise<T> | Thunk<T>): T => {
+	const fiber = fibers.current;
+	if (typeof promiseOrCallback === 'function') {
```

Comment:
TODO f-streams test does not passe. It seems that is when a two waits are done in the same fiber. I am investigating.

Reply:
Fixed, process.nextTick was important in that case.